File: prototype/core/pdfops.py
"""Formatting primitives shared by every content-stream emitter.

Small on purpose. These two conversions were duplicated in drawing.py and
text.py, which is exactly the kind of duplication that drifts: one of the two
starts rounding to five decimals, or stops inverting ink, and the difference
only shows up as a faint smudge on paper.
"""

from __future__ import annotations


def num(value: float) -> str:
    """A PDF number: four decimals, no trailing zeros, no trailing point.

    PDF points are 1/72 inch, so four decimals is ~0.4 micron -- far finer than
    any printer resolves. Short numbers matter because the dot grid is the
    largest stream this program writes.
    """
    text = f"{value:.4f}".rstrip("0").rstrip(".")
    return "0" if text in ("", "-", "-0") else text


def gray(ink: float) -> str:
    """A fraction of black (0.25 = "25 % K") as a PDF gray level.

    PDF gray runs the other way -- 0 is black, 1 is white -- so the inversion
    happens here, once, rather than at every call site.
    """
    if not 0.0 <= ink <= 1.0:
        raise ValueError(f"ink must be a fraction of black in 0..1, got {ink}")
    return num(1.0 - ink)

File: cpp/core/src/pdfops.cc
// pdfops.cc — formatting primitives shared by every content-stream emitter.
//
// Copyright (C) 2026 nava. AGPLv3 or later; see LICENSE.

#include "printbooks/pdfops.hh"

#include <cstdio>
#include <stdexcept>

namespace pb {

std::string num(double value) {
    char buf[64];
    std::snprintf(buf, sizeof buf, "%.4f", value);
    std::string s(buf);

    // Four decimals is ~0.4 micron — far finer than any printer resolves.
    // Short numbers matter because the dot grid is the largest stream written.
    while (!s.empty() && s.back() == '0')
        s.pop_back();
    if (!s.empty() && s.back() == '.')
        s.pop_back();
    if (s.empty() || s == "-" || s == "-0")
        return "0";
    return s;
}

std::string gray(double ink) {
    if (!(ink >= 0.0 && ink <= 1.0))
        throw std::invalid_argument("ink must be a fraction of black in 0..1");
    return num(1.0 - ink);
}

}  // namespace pb
